src/fix_ea/fix_ea.c

```diff
@@ -0,0 +1,555 @@
+
+	if ((buf = malloc(ret)) == NULL) {
+		warn("malloc");
+		return (-1);
+	}
+
+	buflen = ret;
+	if ((ret = extattr_list_fd(fd, EXTATTR_NAMESPACE_USER, buf, buflen)) < 0) {
+		free(buf);
+		warn("extattr_list_fd");
+		return (-1);
+	}
+
+	for (i = 0;i < ret;i += ch + 1) {
+		struct xattr *xptr = NULL;
+		char *name, *value;
+		int getret;
+
+		ch = (unsigned char)buf[i];
+		if ((name = malloc(ch)) == NULL) {
+			err(EX_OSERR, NULL);
+			continue;
+		}
+
+		strncpy(name, &buf[i + 1], ch);
+		name[ch] = '\0';
+
+		if (strncmp(name, "DosStream.", 10) != 0) {
+			free(name);
+			continue;
+		}
+
+		if ((getret = extattr_get_fd(fd, EXTATTR_NAMESPACE_USER,
+			name, NULL, 0)) < 0) {
+			free(name);
+			continue;
+		}
+
+		if ((value = malloc(getret)) == NULL) {
+			warn("malloc");	
+			free(name);
+			continue;
+		}
+
+		if ((getret = extattr_get_fd(fd, EXTATTR_NAMESPACE_USER,
+			name, value, getret)) < 0) {
+			free(value);
+			free(name);
+			continue;
+		}
+
+		if ((xptr = malloc(sizeof(*xptr))) == NULL) {
+			warn("malloc");	
+			free(value);
+			free(name);
+			continue;
+		}
+
+		memset(xptr, 0, sizeof(*xptr));
+		xptr->name = name;
+		xptr->value = value;
+		xptr->length = getret;
+
+		TAILQ_INSERT_TAIL(xlist, xptr, link);
+	}
+
+	return (0);
+}
+
+static int
+get_afp_list(struct xattr_list *xlist, struct xattr_list *afp_list)
+{
+	struct xattr *xptr = NULL;
+
+	if (xlist == NULL || afp_list == NULL)
+		return (-1);
+
+	TAILQ_FOREACH(xptr, xlist, link) {
+		if (xptr->length >= 3 && AFP_EA_CORRUPTED(xptr->value))
+			TAILQ_INSERT_TAIL(afp_list, xptr, afp_link);
+	}
+
+	return (0);
+}
+
+static void
+hexdump_ea(const char *path, const char *name, const char *buf, size_t length)
+{
+	int i;
+
+	if (path == NULL || name == NULL || buf == NULL || length == 0)
+		return;
+
+	printf("%s: %s\n\t", path, name);
+	if (length < 8) {
+		for (i = 0;i < length;i++)
+			printf("%02x ", (unsigned char)buf[i]);
+	} else {
+		for (i = 0;i < 4;i++)
+			printf("%02x ", (unsigned char)buf[i]);
+		printf("/ ");
+		for (i = length - 4;i < length;i++)
+			printf("%02x ", (unsigned char)buf[i]);
+
+	}
+	printf("[%zu]\n", length);
+}
+
+static int
+fix_afp_list(int fd, const char *path,
+		u_int64_t flags, struct xattr_list *afp_list)
+{
+	int ret = 0, setret = 0;
+	struct xattr *xptr = NULL, *xtmp = NULL;
+
+	if (afp_list == NULL)
+		return (-1);
+
+	TAILQ_FOREACH(xptr, afp_list, afp_link) {
+		if (flags & F_DEBUG)
+			hexdump_ea(path, xptr->name, xptr->value, xptr->length);
+
+		if (flags & F_CHECK_AFP_EA) {
+			ret |= EX_EA_CORRUPTED;
+			if (flags & F_VERBOSE) {
+				printf("%s: %s is corrupted\n", path, xptr->name);
+			}
+		}
+
+		if (flags & F_FIX_AFP_EA) {
+			if ((flags & F_DRY_RUN) == 0) {
+				*((char *)xptr->value) = 'A';
+				if ((setret = extattr_set_fd(fd, EXTATTR_NAMESPACE_USER,
+					xptr->name, xptr->value, xptr->length)) < 0) {
+					warn("extattr_set_fd");
+					ret |= EX_EA_CORRUPTED;
+				} 
+			}
+
+			if (setret > 0 || flags & F_DRY_RUN) {
```

Comment:
You can check, as additional measure, that `setret == xptr->length`.

Reply:
This is fine.